### PR TITLE
[SNOW-1413575] Better yaml validation

### PR DESCRIPTION
**Pull request by sfc-gh-dasilva**
Adds `strictyaml` to validate every yaml against a schema derived from the protobuf definition

### CODE REVIEW

File: semantic_model_generator/validate/schema.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+
+def is_optional_field(field_descriptor: FieldDescriptor) -> bool:
+    optional_option = list(
+        filter(
+            lambda o: o[0].name == "optional",
+            field_descriptor.GetOptions().ListFields(),
+        )
+    )
+    return len(optional_option) > 0 and optional_option[0][1]
```
Comment (sfc-gh-jhilgart) [inline semantic_model_generator/validate/schema.py:63]: what is [0][1] indexing into? add a comment

File: semantic_model_generator/data_processing/proto_utils.py [hunk 2]
```diff
@@ -43,67 +43,17 @@ def proto_to_yaml(message: ProtoMsg) -> str:
...
-
-
-def _sample_value_date2str(yaml_dict: Dict[str, Any]) -> None:
-    """
-    helper for _yaml_date2str
-    """
-    for i, sv in enumerate(yaml_dict.get("sample_values", [])):
-        if isinstance(sv, datetime.datetime):
-            yaml_dict["sample_values"][i] = str(sv)
+    parsed_yaml = load(yaml_str, SCHEMA)
```
Comment (sfc-gh-jhilgart) [inline semantic_model_generator/data_processing/proto_utils.py:57]: Is it possible to also include the line number if this fails?

File: semantic_model_generator/validate/schema.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
+from typing import Dict
```
Comment (sfc-gh-jhilgart) [inline semantic_model_generator/validate/schema.py:1]: can you add a readme to this directory, or a top level comment here, that explains what this file is doing and how to modify/update if needed

File: semantic_model_generator/tests/samples/jaffle_shop_missing_required.yaml [hunk 4]
```diff
@@ -0,0 +1,415 @@
+name: jaffle_shop
```
Comment (sfc-gh-jhilgart) [inline semantic_model_generator/tests/samples/jaffle_shop_missing_required.yaml:1]: nite: rename to missing_required_fields.yaml

File: semantic_model_generator/validate/schema.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        filter(
+            lambda o: o[0].name == "optional",
+            field_descriptor.GetOptions().ListFields(),
+        )
+    )
+    return len(optional_option) > 0 and optional_option[0][1]
+
+
+def create_schema_for_enum(
+    enum: EnumDescriptor, precomputed_types: Dict[str, Validator]
```
Comment (sfc-gh-jhilgart) [inline semantic_model_generator/validate/schema.py:67]: can you add docstrings mentioning that this function may be called again downstream hence the `precomputed_types[enum.name] = schema`

File: semantic_model_generator/tests/yaml_to_semantic_model_test.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import os
+
+import pytest
+from strictyaml import YAMLValidationError
+
+from semantic_model_generator.data_processing.proto_utils import yaml_to_semantic_model
```
Comment (sfc-gh-jhilgart) [inline semantic_model_generator/tests/yaml_to_semantic_model_test.py:6]: If the proto definition changes, is there any way to catch that in these tests? 

Maybe we can create a set of expected fields based on the current proto definition and ensure that matches the current compiled proto spec.

Just want to make sure if we're updating the proto we're also updating the strictyaml verification.

**Review by sfc-gh-jhilgart — COMMENTED** [review]
Thank you for adding!

File: semantic_model_generator/data_processing/proto_utils.py (hunk 2 above)
Comment (sfc-gh-dasilva) [reply semantic_model_generator/data_processing/proto_utils.py:57]: `strictyaml` adds line numbers by default in the error messages

File: semantic_model_generator/tests/yaml_to_semantic_model_test.py (hunk 6 above)
Comment (sfc-gh-dasilva) [reply semantic_model_generator/tests/yaml_to_semantic_model_test.py:6]: what do you mean? The way this is implemented we do DFS in the entire proto definition, so any schema changes would be automatically picked up

Comment (sfc-gh-jhilgart) [reply semantic_model_generator/tests/yaml_to_semantic_model_test.py:6]: got it - makes sense

**Review by sfc-gh-jhilgart — APPROVED** [review]

**Pull request merged**